banyand/metadata/schema/etcd.go

```diff
@@ -282,15 +314,104 @@ func (e *etcdSchemaRegistry) UpdateIndexRuleBinding(ctx context.Context, indexRu
 	if err != nil {
 		return errors.Wrap(err, indexRuleBinding.GetMetadata().GetGroup())
 	}
-	return e.update(ctx, g, formatIndexRuleBindingKey(indexRuleBinding.GetMetadata()), indexRuleBinding)
+
+	irs := make(map[string]*databasev1.IndexRule)
+	// ensure the existence of referencing indexRule(s)
+	for _, irName := range indexRuleBinding.GetRules() {
+		ir, innerErr := e.GetIndexRule(ctx, &commonv1.Metadata{
+			Name:  irName,
+			Group: g.GetName(),
+		})
+		if innerErr != nil {
+			return errors.Wrap(innerErr, "index rule does not exist")
+		}
+
+		irs[irName] = ir
+	}
+
+	// sort rules in ASC order
+	sort.Strings(indexRuleBinding.GetRules())
+
+	// try to find the existing instance
+	// save a copy of rule refs if exists
+	var oldRuleRefs []string
+	old, err := e.GetIndexRuleBinding(ctx, &commonv1.Metadata{
+		Name: indexRuleBinding.GetMetadata().GetName(),
+	})
+
+	if err == nil && old != nil {
+		oldRuleRefs = old.GetRules()
+	}
+
+	err = e.update(ctx, g, &resource{
```

Comment:
Why do you tend to update indexrule when updating indexrulebinding? They are both independent objects, don't have to sync with each other.